src/crashlytics/buildToolsJarHelper.ts

```diff
@@ -0,0 +1,73 @@
+import * as fs from "fs-extra";
+import * as os from "os";
+import * as path from "path";
+import * as spawn from "cross-spawn";
+
+import * as downloadUtils from "../downloadUtils";
+import { FirebaseError } from "../error";
+import { logger } from "../logger";
+import * as rimraf from "rimraf";
+import * as utils from "../utils";
+
+const JAR_CACHE_DIR =
+  process.env.FIREBASE_CRASHLYTICS_BUILDTOOLS_PATH ||
+  path.join(os.homedir(), ".cache", "firebase", "crashlytics", "buildtools");
+
+const JAR_VERSION = "2.9.0";
+const JAR_URL = `https://dl.google.com/android/maven2/com/google/firebase/firebase-crashlytics-buildtools/${JAR_VERSION}/firebase-crashlytics-buildtools-${JAR_VERSION}.jar`;
+
+/**
+ * Returns the path to the jar file, downloading it if necessary.
+ */
+export async function fetchBuildtoolsJar(): Promise<string> {
+  if (process.env.LOCAL_JAR) {
+    return process.env.LOCAL_JAR;
+  }
+
+  let jarPath = path.join(JAR_CACHE_DIR, `crashlytics-buildtools-${JAR_VERSION}.jar`);
+  if (fs.existsSync(jarPath)) {
+    logger.debug(`Buildtools Jar already downloaded at ${jarPath}`);
+    return jarPath;
+  }
+
+  // If the Jar cache directory exists, but the jar for the current version
+  // doesn't, then we're running the CLI with a new Jar version and we can
+  // delete the old version.
+  if (fs.existsSync(JAR_CACHE_DIR)) {
+    logger.debug(
+      `Deleting Jar cache at ${JAR_CACHE_DIR} because the CLI was run with a newer Jar version`
+    );
+    rimraf.sync(JAR_CACHE_DIR);
+  }
+  utils.logBullet("Downloading crashlytics-buildtools.jar to " + jarPath);
+  utils.logBullet(
+    "For open source licenses used by this command, look in the META-INF directory in the buildtools.jar file"
+  );
+  const tmpfile = await downloadUtils.downloadToTmp(JAR_URL);
+  fs.mkdirSync(JAR_CACHE_DIR, { recursive: true });
+  fs.copySync(tmpfile, jarPath);
+
+  // If you set LOCAL_JAR to a path it will override the downloaded buildtools.jar
+  if (process.env.LOCAL_JAR) {
+    jarPath = process.env.LOCAL_JAR;
+  }
+  return jarPath;
+}
+
+/**
+ * Helper function to invoke the given set of arguments on the the executable jar
+ */
+export function runBuildtoolsCommand(jarFile: string, args: string[], debug: boolean): void {
+  args = ["-jar", jarFile].concat(args);
```

Comment:
I would _highly_ recommend not modifying the input (since it's passed by reference), but instead creating a new variable for your args.

And, if you want to be real typescripty:

```typescript
const fullArgs = [
  "-jar",
  "jarFile",
  ...args,
  "-clientName",
  "firebase-cli;...",
];
```

Reply:
Thank you! My ts n00bishness is fully exposed here. :)